### PR TITLE
Move protobuf handling of AccessPath structures to ir/proto

### PR DESCRIPTION
**Pull request by bgogul**
Makes further progress towards #118.

Do not merge this until https://github.com/google-research/raksha/pull/133 is merged.

### CODE REVIEW

File: src/ir/selector.h [hunk 1]
```diff
@@ -17,19 +16,11 @@ namespace raksha::ir {
...
-    if (selector.has_field()) {
-      return Selector(FieldSelector(selector.field()));
-    }
-    LOG(FATAL) << "Found a Selector with an unimplemented specific type.";
-  }
-
   explicit Selector(FieldSelector field_selector)
     : specific_selector_(std::move(field_selector)) {}
 
+  std::variant<FieldSelector> variant() const { return specific_selector_; }
```
Comment (markww) [inline src/ir/selector.h:22]: I'm not wild about exposing the `std::variant` to the outside world, as it breaks the feel of this class being like an interface-style abstract class. Could we make the encode method be a `friend` of this class somehow so we didn't have to expose this everywhere?

File: src/ir/access_path_selectors.h [hunk 2]
```diff
@@ -79,13 +63,12 @@ class AccessPathSelectors {
...
   }
 
-  // Fill the selectors of an arcs::AccessPathProto message using the
-  // information in the AccessPathSelectors object.
-  void FillSelectors(arcs::AccessPathProto &access_path_proto) const {
-    for (auto iter = reverse_selectors_.rbegin();
-      iter != reverse_selectors_.rend(); ++iter) {
-      *access_path_proto.add_selectors() = iter->MakeProto();
-    }
+  std::vector<Selector>::const_reverse_iterator begin() const {
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/access_path_selectors.h:67]: I factored this out to #133. There is already a bug #98. I will leave that bug open.

File: src/ir/selector.h [hunk 3]
```diff
@@ -17,19 +16,11 @@ namespace raksha::ir {
...
-    if (selector.has_field()) {
-      return Selector(FieldSelector(selector.field()));
-    }
-    LOG(FATAL) << "Found a Selector with an unimplemented specific type.";
-  }
-
   explicit Selector(FieldSelector field_selector)
     : specific_selector_(std::move(field_selector)) {}
 
+  absl::variant<FieldSelector> variant() const { return specific_selector_; }
```
[the remark this answers is not in the document]
Comment (markww) [reply src/ir/selector.h]: You fixed the first part of this comment, so resolving. See the other, near-duplicate one I made later.

[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/selector.h]: About std::variant. It should be fixed now.

Why are you against exposing the field? Don't you think we will need to create visitors over such variants outside of the class? `Encode` and `Decode` are just one instance, but I think we will add more. Also, isn't it similar to exposing the types of classes through a visitor pattern?

File: src/ir/selector.h (hunk 1 above)
Comment (bgogul) [reply src/ir/selector.h:22]: Why are you against exposing the underlying types? Don't you think we will need to create visitors over such variants outside of the class? Encode and Decode are just one instance, but I think we will add more? 

Also, isn't it similar to exposing the types of classes through a visitor pattern?

File: src/ir/selector.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (markww) [reply src/ir/selector.h]: I'm not against it exactly, I just wanted to see if there was a clean way around exposing it. But as you've mentioned with the visitor pattern, we've kind of externalized the knowledge of the leaf classes anyway, so I think using this getter is fine.

File: src/ir/selector.h (hunk 1 above)
Comment (markww) [reply src/ir/selector.h:22]: See other conversation, this seems fine.

File: src/ir/proto/access_path.cc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  result.set_handle_connection(handle_root.handle_connection_spec_name());
+  return result;
+}
+
+
+//-----------------
+// AccessPath
+
+// We currently only handle the case in which an AccessPathProto is rooted at a
+// (particle_spec, handle_connection), in which case it
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/proto/access_path.cc]: Done.

File: src/ir/access_path_selectors.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/access_path_selectors.h:68]: #133 is merged now.

**Review by markww — APPROVED** [review]

**Pull request merged**